tools/document-pii/src/main.rs

```diff
@@ -0,0 +1,650 @@
+        self.insert_use_statements(vec![enum_name.clone()]);
+        ALL_TYPES
+            .lock()
+            .unwrap()
+            .insert(enum_name, EnumOrStruct::Enum(node.clone()));
+    }
+
+    fn visit_item_use(&mut self, i: &'ast syn::ItemUse) {
+        let use_statements = usetree_to_paths(&i.tree, &self.module_path)
+            .iter()
+            .filter(|s| s.contains("relay"))
+            .cloned()
+            .collect();
+
+        self.insert_use_statements(use_statements);
+    }
+}
+
+#[derive(Ord, PartialOrd, PartialEq, Eq, Hash, Clone, Debug, Default)]
+struct TypeAndField {
+    // Full path of a type. E.g. relay_common::protocol::Event, rather than just 'Event'.
+    qualified_type_name: String,
+    field_name: String,
+}
+
+fn find_rs_files(dir: &PathBuf) -> Vec<std::path::PathBuf> {
+    let walker = WalkDir::new(dir).into_iter();
+    let mut rs_files = Vec::new();
+
+    for entry in walker.filter_map(walkdir::Result::ok) {
+        if !entry.clone().path().to_string_lossy().contains("/src/") {
+            continue;
+        }
+        if entry.file_type().is_file() && entry.path().extension().map_or(false, |ext| ext == "rs")
+        {
+            rs_files.push(entry.into_path());
+        }
+    }
+    rs_files
+}
+
+#[derive(Debug, Clone)]
+enum EnumOrStruct {
+    Struct(ItemStruct),
+    Enum(ItemEnum),
+}
+
+// Need this to wrap lazy_statics in a mutex. everything is single-threaded so it shouldn't matter.
+unsafe impl Send for EnumOrStruct {}
+
+// In practice this is used as a way to keep state between the different methods on the visitor
+// trait, as the predefined trait stops me from adding the parameters I need.
+#[derive(Default)]
+struct TypeVisitor {
+    module_path: String,
+    current_type: String,
+}
+
+impl<'ast> Visit<'ast> for TypeVisitor {
+    fn visit_item_struct(&mut self, node: &'ast ItemStruct) {
+        self.current_type = node.ident.to_string();
+        // This should stop any infinite recursion
+        if !CURRENT_PATH
+            .lock()
+            .unwrap()
+            .iter()
+            .any(|x| x.qualified_type_name == self.current_type)
+        {
+            for field in node.fields.iter() {
+                self.visit_field(field);
+            }
+        }
+    }
+
+    fn visit_item_enum(&mut self, node: &'ast ItemEnum) {
+        self.current_type = node.ident.to_string();
+        if !CURRENT_PATH
+            .lock()
+            .unwrap()
+            .iter()
+            .any(|x| x.qualified_type_name == self.current_type)
+        {
+            for variant in node.variants.iter() {
+                for field in variant.fields.iter() {
+                    self.visit_field(field);
+                }
+            }
+        }
+    }
+
+    fn visit_variant(&mut self, node: &'ast Variant) {
+        for field in node.fields.iter() {
+            self.visit_field(field);
+        }
+    }
+
+    fn visit_field(&mut self, node: &'ast Field) {
```

Comment:
This method is doing several things, could we split it up to make it easier to read and understand?

Reply:
Ok, so the function does 3 things:

1. push and pop to/from the CURRENT_PATH
2. if it's PII then add it to PII_VALUES
3. recursion logic (with the double for-loops)

since both `1` and `2` are just simple statements without any branching, I moved the recursion logic to a separate function, is this what you had in mind?